DexPriceAggregatorUniswapV3/contracts/DexPriceAggregatorUniswapV3.sol

```diff
@@ -0,0 +1,409 @@
+///         The "spot" price is always the last block's ending price.
+contract DexPriceAggregatorUniswapV3 is IDexPriceAggregator, Ownable {
+    address public immutable uniswapV3Factory;
+    address public immutable weth;
+    uint24 public immutable defaultPoolFee;
+
+    mapping(bytes32 => address) public overriddenPoolForRoute;
+
+    constructor(
+        address _uniswapV3Factory,
+        address _weth,
+        uint24 _defaultPoolFee
+    ) {
+        require(
+            _uniswapV3Factory != address(0),
+            "Can't set UniswapV3Factory address as 0"
+        );
+        require(_weth != address(0), "Can't set WETH address as 0");
+        uniswapV3Factory = _uniswapV3Factory;
+        weth = _weth;
+        defaultPoolFee = _defaultPoolFee;
+    }
+
+    /********************
+     * Oracle functions *
+     ********************/
+
+    /// @inheritdoc IDexPriceAggregator
+    function assetToAsset(
+        address _tokenIn,
+        uint256 _amountIn,
+        address _tokenOut,
+        uint256 _twapPeriod
+    ) external view override returns (uint256 amountOut) {
+        if (_tokenIn == weth) {
+            return ethToAsset(_amountIn, _tokenOut, _twapPeriod);
+        } else if (_tokenOut == weth) {
+            return assetToEth(_tokenIn, _amountIn, _twapPeriod);
+        } else {
+            return
+                _fetchAmountCrossingPools(
+                    _tokenIn,
+                    _amountIn,
+                    _tokenOut,
+                    _twapPeriod
+                );
+        }
+    }
+
+    /// @notice Given a token and its amount, return the equivalent value in ETH
+    /// @param _tokenIn Address of an ERC20 token contract to be converted
+    /// @param _amountIn Amount of tokenIn to be converted
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return ethAmountOut Amount of ETH received for amountIn of tokenIn
+    function assetToEth(
+        address _tokenIn,
+        uint256 _amountIn,
+        uint256 _twapPeriod
+    ) public view returns (uint256 ethAmountOut) {
+        address tokenOut = weth;
+        address pool = _getPoolForRoute(
+            PoolAddress.getPoolKey(_tokenIn, tokenOut, defaultPoolFee)
+        );
+        return
+            _fetchAmountFromSinglePool(
+                _tokenIn,
+                _amountIn,
+                tokenOut,
+                pool,
+                _twapPeriod
+            );
+    }
+
+    /// @notice Given an amount of ETH, return the equivalent value in another token
+    /// @param _ethAmountIn Amount of ETH to be converted
+    /// @param _tokenOut Address of an ERC20 token contract to convert into
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return amountOut Amount of tokenOut received for ethAmountIn of ETH
+    function ethToAsset(
```

Comment:
- _ethAmountIn should not be zero
-  _tokenOut should not be the zero address
- _twapPeriod should be greater than 0

libraries/OracleLibrary.sol

function consult(address pool, uint32 period)
internal
view
returns (int24 timeWeightedAverageTick)
{
require(period != 0, "BP"); ///// -----------> this